src/python/pants/backend/url_handlers/s3/register.py

```diff
@@ -107,40 +109,75 @@ class S3DownloadFile:
 
 @rule
 async def download_from_s3(
-    request: S3DownloadFile, aws_credentials: AWSCredentials, global_options: GlobalOptions
+    request: S3DownloadFile,
+    aws_credentials: AWSCredentials,
+    global_options: GlobalOptions,
+    s3_subsystem: S3Subsystem,
 ) -> Digest:
     from botocore import auth, compat, exceptions  # pants: no-infer-dep
 
-    # NB: The URL for auth is expected to be in path-style
-    path_style_url = "https://s3"
+    virtual_hosted_url = f"https://{request.bucket}.s3.amazonaws.com/{request.key}"
     if request.region:
-        path_style_url += f".{request.region}"
-    path_style_url += f".amazonaws.com/{request.bucket}/{request.key}"
+        virtual_hosted_url = (
+            f"https://{request.bucket}.s3.{request.region}.amazonaws.com/{request.key}"
+        )
     if request.query:
-        path_style_url += f"?{request.query}"
+        virtual_hosted_url += f"?{request.query}"
 
     headers = compat.HTTPHeaders()
-    http_request = SimpleNamespace(
-        url=path_style_url,
-        headers=headers,
-        method="GET",
-        auth_path=None,
-    )
+    signer = None
+    http_request = None
+
+    if s3_subsystem.auth_signing == S3AuthSigning.SIGV4:
+        # sigv4 uses the virtual_hosted_url for the auth request
+        http_request = SimpleNamespace(
+            url=virtual_hosted_url,
+            headers=headers,
+            method="GET",
+            auth_path=None,
+            data=None,
+            params={},
+            context={},
+            body={},
+        )
+
+        # Add x-amz-content-SHA256 as per boto code
+        # ref link - https://github.com/boto/botocore/blob/547b20801770c8ea4255ee9c3b809fea6b9f6bc4/botocore/auth.py#L52C1-L54C2
+        headers.add_header(
+            "X-Amz-Content-SHA256",
+            EMPTY_FILE_DIGEST.fingerprint,
+        )
+
+        # A region is required to sign the request with sigv4. If we don't know where the bucket is,
+        # default to the region from the credentials and fallback to us-east-1
+        signing_region = request.region or aws_credentials.default_region or "us-east-1"
```

Comment:
Could/should the `us-east-1` Pants-default region also be an option on the subsystem?

Reply:
I'm torn on how to handle this one, if I expose it as a subsystem option I'd expect that setting it would have high precedence, but the current logic only uses it as the third and least important choice.

So I could either 
- try and make it clear through naming this is a fallback value 
- make the subsystem option higher priority (first?)
- throw an error instead of having the fallback to us-east-1

Reply:
updated